### ISSUE TITLE
Openvpn GUI crashes. The connection remains.

### ISSUE THREAD

**Issue by MDcod** [OpenVPN/openvpn-gui#548]
Drops randomly. Didn't notice the patterns. The tray icon disappears and the OpenVPN GUI for Windows process disappears.
Network connectivity remains good. But now there is no way to break the connection. Just restart the computer. I don't see gui logs anywhere, only the network itself.

> OS: Windows 11 Pro 21H2 (22000.1219), also crashed on earlier versions
> OpenVpn: (v2.5.8), also crashed on earlier versions
> OpenVpn GUI: (v11.31.0.0), also crashed on earlier versions

Please help, it's incredibly annoying

**Comment by MDcod** [comment]
Maybe I can give some more useful information about it?

**Comment by selvanair** [comment]
> Maybe I can give some more useful information about it?

Yes please !

**Comment by selvanair** [comment]
OpenVPN logs may help for a start. When the GUI crashes leaving openvpn.exe running, go to `OpenVPN\log` in your user profile folder and look for a file named `<foo>.log`  where `<foo>` is the name of the connection that was started. Also look into Windows event logs.

Alternatively, when the GUI crashes, got to task-manager and kill the openvpn.exe process left behind. That will terminate the VPN connection -- no need to reboot. Then restart the GUI and from its menu open the log file of the connection that crashed (connection-name->View Log). Upload it somewhere and post a link here.

Another thing to do is to narrow down the issue to a particular OpenVPN/OS version combination. Your initial post says
 
> OS: Windows 11 Pro 21H2 (22000.1219), also crashed on earlier versions
> OpenVpn: (v2.5.8), also crashed on earlier versions
> OpenVpn GUI: (v11.31.0.0), also crashed on earlier versions

That's not helpful. If every version is crashing on all versions of the OS, what can I say.

**Comment by mduivcap** [comment]
I'm having the same issue as the poster mentioned but with a newer version.

The app does crash when I try to connect en when I try to disconnect.
In between it usually does not crash.
Unfortunatelly I cannot end the task (company policy).

The logfile:

> 2023-03-10 15:10:46 WARNING: Compression for receiving enabled. Compression has been used in the past to break encryption. Sent packets are not compressed unless "allow-compression yes" is also set.
2023-03-10 15:10:46 Note: --cipher is not set. OpenVPN versions before 2.5 defaulted to BF-CBC as fallback when cipher negotiation failed in this case. If you need this fallback please add '--data-ciphers-fallback BF-CBC' to your configuration and/or add BF-CBC to --data-ciphers.
2023-03-10 15:10:46 Note: '--allow-compression' is not set to 'no', disabling data channel offload.
2023-03-10 15:10:46 OpenVPN 2.6.0 [git:v2.6.0/b999466418dddb89] Windows-MSVC [SSL (OpenSSL)] [LZO] [LZ4] [PKCS11] [AEAD] [DCO] built on Jan 25 2023
2023-03-10 15:10:46 Windows version 10.0 (Windows 10 or greater), amd64 executable
2023-03-10 15:10:46 library versions: OpenSSL 3.0.7 1 Nov 2022, LZO 2.10
2023-03-10 15:10:46 MANAGEMENT: TCP Socket listening on [AF_INET]127.0.0.1:25340
2023-03-10 15:10:46 Need hold release from management interface, waiting...
2023-03-10 15:10:46 MANAGEMENT: Client connected from [AF_INET]127.0.0.1:50229
2023-03-10 15:10:47 MANAGEMENT: CMD 'state on'
2023-03-10 15:10:47 MANAGEMENT: CMD 'log on all'
2023-03-10 15:10:47 MANAGEMENT: CMD 'echo on all'
2023-03-10 15:10:47 MANAGEMENT: CMD 'bytecount 5'
2023-03-10 15:10:47 MANAGEMENT: CMD 'state'
2023-03-10 15:10:47 MANAGEMENT: CMD 'hold off'
2023-03-10 15:10:47 MANAGEMENT: CMD 'hold release'
2023-03-10 15:13:17 MANAGEMENT: CMD 'username "Auth" "username"'
2023-03-10 15:13:18 MANAGEMENT: Client disconnected

The weird part is that even when the connection is established, this logfile does not change, it will still say Client disconnected at the end. I know VPN is working when I can reach a certain url.

When you startup openvpngui again, it does not show that there is a connection up and running, so besides the fact that the gui should not crash, it would be nice to at least be able to start it up again and see if there are connections.

Is there anything else we can provide to make this problem more clear?

**Comment by mduivcap** [comment]
Maybe it was resolved with a Windows update, but after I posted this log, it didn't crash anymore (while before it would crash no matter what I did).

**Comment by mduivcap** [comment]
Could this be the answer?
https://answers.microsoft.com/en-us/windows/forum/all/app-crash-with-ntdlldll/9aa59f80-99a1-4f3c-b2f0-7eb72df05b2a

**Comment by selvanair** [comment]
Faulted in ntdll.dll does not mean that the error is there. It could still be the application that is causing the dll to crash -- it just makes debugging harder.

If you can provide exact commands you use, any other actions you do that could affect the config files between connect and the crash, I can try to reproduce.

>2023-03-10 15:10:47 MANAGEMENT: CMD 'hold release'
2023-03-10 15:13:17 MANAGEMENT: CMD 'username "Auth" "username"'
2023-03-10 15:13:18 MANAGEMENT: Client disconnected

Assuming that "disconnect" is  because the GUI crashed, does it always happen after sending username as in the log snippet above?

**Comment by mduivcap** [comment]
The exact commands I don't know. 
**Connect > Crash**
What I do in the GUI is, I fill in the password and hit "OK"
The crash happens when the connection seems to succeed. Sometimes I see the screen in the picture just disappear (no Windows notification) and when I go to the system tray icon, it's there, but when I hover over it, it disappears (because it crashed). 
Sometimes it's similar, but I do briefly see the Windows notification popping up
And as mentioned, sometimes it does not crash and I see the notification and the system tray Icon remains where it is as it should
![image](https://user-images.githubusercontent.com/127507539/225254083-d67953b9-4424-4544-8af7-313c3c8fbebc.png)
![image](https://user-images.githubusercontent.com/127507539/225255121-1ed11ab3-ebda-494d-b37a-d363cbe5109b.png)
**Connect > Crash**
What I do is open the system tray, rightclick on the icon, click on the connection that is active, click disconnect.
It seems to crash at the moment of disconnecting (when I have the log open, I do see something happening before it crashes and afterwards I'm not connected anymore, so it did disconnect, it just also made the app crash)
![image](https://user-images.githubusercontent.com/127507539/225255833-e5f9ffe3-ca4a-4e9c-9abb-efe12e766ba8.png)

**Comment by selvanair** [comment]
> The exact commands I don't know.

I misunderstood that you were sending commands to a running GUI as in a previous crash report on a related thread. 

Thanks for the details -- looks like very common and normal usage. The connection remaining after GUI crash is expected as OpenVPN.exe will continue to run. Also, restarting the GUI in such cases will not work right, as it wont be able to connect back to the running OpeVPN.exe process.

But the GUI should not crash in the first place. We'll see whether anyone with access to Windows 11 Pro can reproduce this: ping @lstipakov

**Comment by lstipakov** [comment]
I have Windows 11 (ARM64, not sure if that matters) but was unable to reproduce this, even though I use gui somewhat actively. Will keep an eye on it.

**Comment by mduivcap** [comment]
A, good to mention indeed. I'm using Win11 on x64

**Comment by WigF1** [comment]
> > The exact commands I don't know.
> 
> I misunderstood that you were sending commands to a running GUI as in a previous crash report on a related thread.
> 
> Thanks for the details -- looks like very common and normal usage. The connection remaining after GUI crash is expected as OpenVPN.exe will continue to run. Also, restarting the GUI in such cases will not work right, as it wont be able to connect back to the running OpeVPN.exe process.
> 
> But the GUI should not crash in the first place. We'll see whether anyone with access to Windows 11 Pro can reproduce this: 

I use Win11 Pro per 

![image](https://user-images.githubusercontent.com/12164329/228432318-da4e9b09-ea0a-4ded-baf1-11227877d93f.png)

I have this exact issue, I've had it on Win11 21H2 and Win11 22H2. It can sometimes be 4,5, or 6 attempts before I am able to establish a connection. 

I usually end up restarting the OpenVPN Interactive Service, then re-opening the GUI. 

Based on the Exception code that appears in the Application event log, it would seem to suggest something other than end-user environments:

https://learn.microsoft.com/en-us/answers/questions/1001963/ntdll-dll-application-error-(0xc0000374)

![image](https://user-images.githubusercontent.com/12164329/228433092-f6ba2478-29f8-486a-86dc-8fcdb530b5a9.png)

**Comment by chironh** [comment]
Having the same issue on Windows 11 22H2 with both 2.6.3 and 2.6.4.

OpenVPN icon just disappears from taskbar, but VPN connection is still active. Only way to kill the connection is to restart the OpenVPN Interactive Service, as mentioned by [WigF1](https://github.com/WigF1).

`Faulting application name: openvpn-gui.exe, version: 11.42.0.0, time stamp: 0x645cb914
Faulting module name: ntdll.dll, version: 10.0.22621.1485, time stamp: 0xb7f0154a
Exception code: 0xc0000374
Fault offset: 0x000000000010be19
Faulting process id: 0x0xBF4
Faulting application start time: 0x0x1D98C903E1BDAC1
Faulting application path: C:\Program Files\OpenVPN\bin\openvpn-gui.exe
Faulting module path: C:\Windows\SYSTEM32\ntdll.dll
Report Id: 3e595943-d88b-4795-8fae-20cd0d0d7bf2
Faulting package full name: 
Faulting package-relative application ID: `

**Comment by chironh** [comment]
> Do you have any specific steps to reproduce it?

Not really, it seems it just crashes while using. 
Most of the time I only notice the crash when I want to disconnect and the icon's missing.

Is there a setting to enable verbose logs?

**Comment by mduivcap** [comment]
> Do you have any specific steps to reproduce it?

It mainly happens when trying to connect (does not matter if it passes or fails) and when you try to disconnect.
I might have seen it crash while it was connected, but I'm not fully sure anymore.

**Comment by lstipakov** [comment]
I have attached `openvpn-gui.exe` with debug symbols enabled, `openvpn-gui.pdb` which contains debug symbols and registry file to enable dump collection for `openvpn-gui.exe`.

Could you please:

 1. Enable dump collection for `openvpn-gui.exe` by running `.\enable-gui-dump.reg` under administrative privileges. This should create key `HKLM\SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps\openvpn-gui.exe`.
 2. Copy `openvpn-gui.exe` to `C:\Program Files\OpenVPN\bin`.
 3. If openvpn-gui is running, close it and rerun.
 4. Reproduce the crash.
 5. A crash dump should appear at `%LOCALAPPDATA%\CrashDumps`. Paste it here.


[gui-dump.zip](https://github.com/OpenVPN/openvpn-gui/files/11563763/gui-dump.zip)

**Comment by Alexey-Kovalenko** [comment]
Well, I'm not any of above speakers, but have same issue.
Openvpn-gui randomly disappear (silently crashes).
My keys are password protected.
Here is a dump.
[openvpn-gui.exe.7568.zip](https://github.com/OpenVPN/openvpn-gui/files/11584950/openvpn-gui.exe.7568.zip)

**Comment by lstipakov** [comment]
> Here is a dump

Thanks. Sadly it is not super clear from the dump where is the problem, apart from "heap corruption".

```
Microsoft (R) Windows Debugger Version 10.0.22621.755 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.


Loading Dump File [C:\Temp\dump\gui\openvpn-gui.exe.7568.dmp]
User Mini Dump File: Only registers, stack and portions of memory are available

Symbol search path is: srv*
Executable search path is: 
Windows 10 Version 19045 MP (4 procs) Free x64
Product: WinNt, suite: SingleUserTS
Edition build lab: 19041.1.amd64fre.vb_release.191206-1406
Machine Name:
Debug session time: Sun May 28 18:01:04.000 2023 (UTC + 3:00)
System Uptime: not available
Process Uptime: 0 days 0:01:10.000
.........................................................
This dump file has an exception of interest stored in it.
The stored exception information can be accessed via .ecxr.
(1d90.abc): Unknown exception - code c0000374 (first/second chance not available)
For analysis of this file, run !analyze -v
ntdll!NtWaitForMultipleObjects+0x14:
00007ffe`6366dbb4 c3              ret
0:003> !analyze -v
*******************************************************************************
*                                                                             *
*                        Exception Analysis                                   *
*                                                                             *
*******************************************************************************

*** WARNING: Unable to verify timestamp for openvpn-gui.exe

KEY_VALUES_STRING: 1

    Key  : Analysis.CPU.mSec
    Value: 2937

    Key  : Analysis.DebugAnalysisManager
    Value: Create

    Key  : Analysis.Elapsed.mSec
    Value: 88931

    Key  : Analysis.Init.CPU.mSec
    Value: 3468

    Key  : Analysis.Init.Elapsed.mSec
    Value: 21500

    Key  : Analysis.Memory.CommitPeak.Mb
    Value: 129

    Key  : Timeline.Process.Start.DeltaSec
    Value: 70

    Key  : WER.OS.Branch
    Value: vb_release

    Key  : WER.OS.Timestamp
    Value: 2019-12-06T14:06:00Z

    Key  : WER.OS.Version
    Value: 10.0.19041.1

    Key  : WER.Process.Version
    Value: 11.42.0.0


FILE_IN_CAB:  openvpn-gui.exe.7568.dmp

NTGLOBALFLAG:  0

PROCESS_BAM_CURRENT_THROTTLED: 0

PROCESS_BAM_PREVIOUS_THROTTLED: 0

APPLICATION_VERIFIER_FLAGS:  0

CONTEXT:  (.ecxr)
rax=00007ffe63744a24 rbx=00000000c0000374 rcx=0000001389efeef0
rdx=0000001389eff3f0 rsi=0000000000000001 rdi=00007ffe637397f0
rip=00007ffe636cf449 rsp=0000001389eff3f0 rbp=0000000000000000
 r8=0000001389eff400  r9=0000000000000000 r10=0000000000000202
r11=0000001389efeef0 r12=0000000000000000 r13=0000013f819e9d50
r14=0000013f819e9d40 r15=0000000000000001
iopl=0         nv up ei pl nz na po nc
cs=0033  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000206
ntdll!RtlReportFatalFailure+0x9:
00007ffe`636cf449 eb00            jmp     ntdll!RtlReportFatalFailure+0xb (00007ffe`636cf44b)
Resetting default scope

EXCEPTION_RECORD:  (.exr -1)
ExceptionAddress: 00007ffe636cf449 (ntdll!RtlReportFatalFailure+0x0000000000000009)
   ExceptionCode: c0000374
  ExceptionFlags: 00000001
NumberParameters: 1
   Parameter[0]: 00007ffe637397f0

PROCESS_NAME:  openvpn-gui.exe

ERROR_CODE: (NTSTATUS) 0xc0000374 - A heap has been corrupted.

EXCEPTION_CODE_STR:  c0000374

EXCEPTION_PARAMETER1:  00007ffe637397f0

STACK_TEXT:  
00000013`89eff3f0 00007ffe`636cf413     : 00007ffe`6371760c 00000013`89eff8f0 00000000`00000000 0000013f`819e9050 : ntdll!RtlReportFatalFailure+0x9
00000013`89eff440 00007ffe`636d8192     : 00000000`00000001 00007ffe`637397f0 00000000`00000003 0000013f`81970000 : ntdll!RtlReportCriticalFailure+0x97
00000013`89eff530 00007ffe`636d847a     : 00000000`00000003 00000000`00000000 0000013f`81970000 00000000`00000003 : ntdll!RtlpHeapHandleError+0x12
00000013`89eff560 00007ffe`636de101     : 0000013f`81970000 0000013f`81970000 00007ffe`61052c80 00007ffe`60f79af3 : ntdll!RtlpHpHeapHandleError+0x7a
00000013`89eff590 00007ffe`63677482     : 00000000`00000003 00007ffe`62f64449 00000000`00000001 00007ffe`635e9877 : ntdll!RtlpLogHeapFailure+0x45
00000013`89eff5c0 00007ffe`635f47b1     : 00000000`7ffe0385 0000013f`81970000 0000013f`81970000 00000000`00000000 : ntdll!RtlpFreeHeapInternal+0x81d72
00000013`89eff680 00007ffe`6364053c     : 0000013f`819e36d8 00000000`0000000d 00000013`8974c000 00000000`00000000 : ntdll!RtlFreeHeap+0x51
00000013`89eff6c0 00007ffe`635e7665     : 00000013`8974c000 00000000`00000000 00000000`00000000 0000013f`819730c0 : ntdll!LdrpFreeTls+0x84
00000013`89eff6f0 00007ffe`6362467e     : 0000013f`819d2100 00000000`00000000 00000000`00000000 00000000`00000000 : ntdll!LdrShutdownThread+0x215
00000013`89eff7f0 00007ffe`6121f3aa     : 00007ffe`60390000 00000000`00000000 0000013f`819d21c0 00000000`00000000 : ntdll!RtlExitUserThread+0x3e
00000013`89eff830 00007ffe`603a20df     : 00000000`00000000 00000000`00000001 00000000`00000001 00007ffe`60390000 : KERNELBASE!FreeLibraryAndExitThread+0x4a
00000013`89eff860 00007ffe`63317614     : ffffffff`ffffffff ffffffff`ffffffff 00000000`00000000 00000000`00000000 : mswsock!SockAsyncThread+0x110cf
00000013`89eff8c0 00007ffe`636226a1     : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : kernel32!BaseThreadInitThunk+0x14
00000013`89eff8f0 00000000`00000000     : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : ntdll!RtlUserThreadStart+0x21


STACK_COMMAND:  ~3s; .ecxr ; kb

SYMBOL_NAME:  mswsock!SockAsyncThread+110cf

MODULE_NAME: mswsock

IMAGE_NAME:  mswsock.dll

FAILURE_BUCKET_ID:  HEAP_CORRUPTION_c0000374_mswsock.dll!SockAsyncThread

OS_VERSION:  10.0.19041.1

BUILDLAB_STR:  vb_release

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

IMAGE_VERSION:  10.0.19041.546

FAILURE_ID_HASH:  {0e6358ef-db41-8e79-4ff7-0745ef881149}

Followup:     MachineOwner
---------

```

**Comment by Alexey-Kovalenko** [comment]
At least, I think bug in TCP/IP communications with openvpn.exe, started by service (HEAP_CORRUPTION_c0000374_mswsock.dll!SockAsyncThread)
[CrashDumps.zip](https://github.com/OpenVPN/openvpn-gui/files/11592770/CrashDumps.zip)
I made two more dumps, running procmon in parallel.

**Comment by lstipakov** [comment]
The second dump looks more interesting:

```
00 00000065`323fca28 00007ffe`636add88     ntdll!NtWaitForMultipleObjects+0x14
01 00000065`323fca30 00007ffe`636ad36e     ntdll!WerpWaitForCrashReporting+0xa8
02 00000065`323fcab0 00007ffe`636acb2b     ntdll!RtlReportExceptionHelper+0x33e
03 00000065`323fcb80 00007ffe`636cf484     ntdll!RtlReportException+0x9b
04 00000065`323fcc00 00007ffe`6365c876     ntdll!RtlReportFatalFailure$filt$0+0x33
05 00000065`323fcc30 00007ffe`636723df     ntdll!_C_specific_handler+0x96
06 00000065`323fcca0 00007ffe`636214a4     ntdll!RtlpExecuteHandlerForException+0xf
07 00000065`323fccd0 00007ffe`636211f5     ntdll!RtlDispatchException+0x244
08 00000065`323fd3e0 00007ffe`636cf449     ntdll!RtlRaiseException+0x185
09 00000065`323fda60 00007ffe`636cf413     ntdll!RtlReportFatalFailure+0x9
0a 00000065`323fdab0 00007ffe`636d8192     ntdll!RtlReportCriticalFailure+0x97
0b 00000065`323fdba0 00007ffe`636d847a     ntdll!RtlpHeapHandleError+0x12
0c 00000065`323fdbd0 00007ffe`636de101     ntdll!RtlpHpHeapHandleError+0x7a
0d 00000065`323fdc00 00007ffe`63677482     ntdll!RtlpLogHeapFailure+0x45
0e 00000065`323fdc30 00007ffe`635f47b1     ntdll!RtlpFreeHeapInternal+0x81d72
0f 00000065`323fdcf0 00007ffe`2d322248     ntdll!RtlFreeHeap+0x51
10 00000065`323fdd30 00007ffe`2d2c6a50     riched20!CW32System::FreePv+0x1c
11 00000065`323fdd60 00007ffe`2d313030     riched20!CDisplay::Thaw+0xb4
12 00000065`323fde50 00007ffe`2d2b38c1     riched20!CTxtEdit::SetText+0x6d4
13 00000065`323fe030 00007ffe`2d2b630b     riched20!CTxtEdit::TxSendMessage+0xa41
14 00000065`323fe2e0 00007ffe`6258e858     riched20!RichEditWndProc+0xbeb
15 00000065`323fece0 00007ffe`6258de1b     user32!UserCallWinProcCheckWow+0x2f8
16 00000065`323fee70 00007ffe`6258d68a     user32!SendMessageWorker+0x70b
17 00000065`323fef10 00007ff6`3d7cae07     user32!SendMessageW+0xda
18 00000065`323fef60 00000174`8f49d3ad     openvpn_gui!OnLogLine+0x227 [C:\Users\lev\Projects\openvpn-build\src\openvpn-gui\openvpn.c @ 223] 
19 00000065`323fef68 00000000`00000001     0x00000174`8f49d3ad
1a 00000065`323fef70 00000174`8f449e90     0x1
1b 00000065`323fef78 00007ffe`635f47b1     0x00000174`8f449e90
1c 00000065`323fef80 00000000`00000000     ntdll!RtlFreeHeap+0x51

```

**Comment by selvanair** [comment]
Being a heap corruption, its hard to say where the fault really is. It could be anywhere before the crash triggers, and the exception location in the dumps do not look helpful. Is there something like valgrind on Windows that could check invalid reads and writes etc?

**Comment by chipitsine** [comment]
I like that tool https://github.com/dynamorio/drmemory

**Comment by mduivcap** [comment]
For what's it worth. I do need to connect to two VPN's and I noticed that with one connection the openvpngui app crashes more often than the other. I'll ask if I'm allowed to share some details about the differences between the two, maybe this gives a hint in where (part of) the problem might be.

**Comment by sscotter** [comment]
I have around 130 users which connect to our company network using OpenVPN and the OpenVPN GUI. One of those users is regularly having problems with the GUI silently crashing (and as others mention, the VPN stays up.. just the GUI disappears from the system tray). From what I gather it occurs most days, multiple times a day.. but occasionally it'll be fine for a day, then revert to crashing the following day. There's no rhythm nor reason to it. Nothing out of the usual with his machine or his use case compared to any of his colleagues.

The issue was first reported in April 2023 and we've tried the following
* Upgrading to the latest versions are soon as they come out
* Compared and contrasted his machine to his colleagues 
* Deleting and recreating his account on the firewall, which results in creating him a new VPN profile (including certificate and key)
* Reinstalled Windows
* Multiple other things I do not recall!

He is part of a team which normally connects to two different VPNs simultaneously. There is around 60 people in that team and no one else has the same issue. He feels the GUI crashes more often than not when connecting to one particular VPN.

When the problem persisted across windows reinstallations I felt the problem must be related to his VPN configuration file but I've since borrowed his VPN config and ran it on my machine over the weekend and didn't have any problems.

Very frustrating and confusing.

If anyone has any suggestions on how to fix this OR how to possibly gain more diagnostic information to try and get to the root case I'd be grateful!

Thanks!

**Comment by selvanair** [comment]
> If anyone has any suggestions on how to fix this OR how to possibly gain more diagnostic information to try and get to the root case I'd be grateful!

Most helpful feedback would be one that could help us reproduce the crash. Would you happen to know from which version of the GUI this started? So far crash dumps have not been particularly useful as it seems to be a heap corruption which is hard to pin down. Especially so when this happens only on some end-user installations.

**Comment by lstipakov** [comment]
Since 2.6.5 we should have dump collection [feature](https://github.com/OpenVPN/openvpn-build/commit/b72896baa11a7845944622fe68f9d3786e914905). @sscotter Can you check if there is anything in `%LOCALAPPDATA%\CrashDumps` ?

**Comment by sscotter** [comment]
@selvanair  .. looking back at my conversations with the user I can see they first reported the issue on 9th February 2023 (not April 2023 as I previously stated). Looking at the software inventory history of the machine it appears it was upgraded from 2.5.8-I604 to 2.6.0-I003 on the 8th at my request for compliance reasons. We can see he went back to 2.5.8 on 13th Feb for two days, presumable in an attempt to restore stability, which I suspect was fruitless because he went back to 2.6.0 on the 15th.

Full timeline of changes of version installed on the machine
```
Dec 15th 2022
Uninstalled  Software  [OpenVPN 2.5.5-I602 amd64]
Installed Software  [OpenVPN 2.5.8-I604 amd64]

Feb 8th 2023
Uninstalled  Software  [OpenVPN 2.5.8-I604 amd64]
Installed Software  [OpenVPN 2.6.0-I003 amd64]

**** PROBLEM REPORTED HERE **** 

Feb 10th 2023
Uninstalled  Software  [OpenVPN 2.6.0-I003 amd64]
Installed Software  [OpenVPN 2.6.0-I004 amd64]

Feb 13th 2023
Installed Software  [OpenVPN 2.5.8-I604 amd64]
Uninstalled  Software  [OpenVPN 2.6.0-I004 amd64]

Feb 15th 2023
Uninstalled  Software  [OpenVPN 2.5.8-I604 amd64]
Installed Software  [OpenVPN 2.6.0-I003 amd64]

Mar 16th 2023
Uninstalled  Software  [OpenVPN 2.6.0-I003 amd64]
Installed Software  [OpenVPN 2.6.1-I001 amd64]

Mar 30th 2023
Uninstalled  Software  [OpenVPN 2.6.1-I001 amd64]
Installed Software  [OpenVPN 2.6.2-I001 amd64]

Apr 19th 2023
Uninstalled  Software  [OpenVPN 2.6.2-I001 amd64]
Installed Software  [OpenVPN 2.6.3-I001 amd64]

Jul 27th 2023
Uninstalled  Software  [OpenVPN 2.6.3-I001 amd64]
Installed Software  [OpenVPN 2.6.5-I001 amd64]
```

We've upgraded the machine to 2.6.6 this morning.. unfortunately the problem persists.

---

@lstipakov in regards to your request for the Crash Dumps, we've checked the machine and there is no `AppData\Local\CrashDumps` directory in the users home directory. There is a CrashDump for another local admin account used to installer software though, but that only contains dumps for msiexec.exe and not related to OpenVPN in any way.

```
Checking C:\Users\USER\AppData\Local\CrashDumps...
Path does not exist. No crashdumps available.
Checking C:\Users\LOCALADMIN\AppData\Local\CrashDumps...
Path exists! Heres a list of all the crashdumps available.


    Directory: C:\Users\LOCALADMIN\AppData\Local\CrashDumps


Mode                 LastWriteTime         Length Name                                                                 
----                 -------------         ------ ----                                                                 
-a----        26/07/2023     11:45        2993585 msiexec.exe.4328.dmp                                                 
-a----        26/07/2023     11:42        2995437 msiexec.exe.9580.dmp                                                 
Checking C:\Users\Public\AppData\Local\CrashDumps...
Path does not exist. No crashdumps available.
```

We checked the machines registry, it had keys for `openvpn.exe`, `openvpn-gui.exe` and `openvpnserv.exe` in `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps`

---
@selvanair and @lstipakov, thank you both for taking the time to respond. It's appreciated!

**Comment by lstipakov** [comment]
One way would be to `git bisect` between those releases and pinpoint the exact commit which causes this issue. 

```
PS C:\Users\lev\Projects\openvpn-build\src\openvpn-gui> git bisect bad
Bisecting: 30 revisions left to test after this (roughly 5 steps)
[8b1976c6e35833ec02ee15deba440454cc888283] Increase max size of management password buffer
```

@sscotter do you think it would be feasible for you to test ~5 gui executables one by one? I can of course provide binaries.

**Comment by lstipakov** [comment]
Let's start with this one, built from commit [8b1976c](https://github.com/OpenVPN/openvpn-gui/commit/8b1976c6e35833ec02ee15deba440454cc888283).
[openvpn-gui-8b1976c.zip](https://github.com/OpenVPN/openvpn-gui/files/12358699/openvpn-gui-8b1976c.zip)

**Comment by sscotter** [comment]
Hi @lstipakov,

Just to give you an update.. we've deployed the 8b1976c build this morning and so far it's behaving. However it's not unheard for it to behave for a day or two, then revert to having regular problems again. We'll let you know quickly if it crashes again, but if you don't hear from us for a day or two it probably means it's behaving.

To put a slight spanner in the works I'm on vacation from tomorrow for a week. If the user is still having problems I'll post their Github handle here so you two can continue to troubleshoot in my absence.

Again, thanks for taking the time to reply and help us troubleshoot this issue.

**Comment by sscotter** [comment]
The user reported the GUI behaved all day today. I may be because the problem is resolved.. it may be because he was lucky and I'll come back with a vengeance tomorrow. Lets hope it's for the former.

The user has signed up to github today and his handle is j4ypeem. It you receive any messages from him, its to do with this. Thanks1

**Comment by lstipakov** [comment]
@sscotter Unfortunately I cannot tag @j4ypeem here. If the version https://github.com/OpenVPN/openvpn-gui/commit/8b1976c6e35833ec02ee15deba440454cc888283 still works, I would like to continue with bisecting and provide the next build.

**Comment by j4ypeem** [comment]
@lstipakov Thanks for your help with this issue. Since last week I've been running the build you shared with no problems whatsoever.

I noticed there was a comment on that build regarding max size of password. Is there a known password character limit? Would a 30 char password be a likely factor in this issue?

I'm happy to try another version if you want to troubleshoot further.

**Comment by lstipakov** [comment]
Hi @j4ypeem, 

Thanks for the testing, we're moving forward. Here is the next version from commit https://github.com/OpenVPN/openvpn-gui/commit/c354c19e8fbdd193fbdf212af9c74f3b33a65025. Please give it a try. After this one we'll have only 4 steps left to find offensive commit.

[openvpn-gui-c354c1.zip](https://github.com/OpenVPN/openvpn-gui/files/12407975/openvpn-gui-c354c1.zip)

Note that the password mentioned in the commit message is the management interface password, which is normally set by openvpn-gui and not something user controls, except persistent connections / PLAP cases. Anyway, starting from commit https://github.com/OpenVPN/openvpn-gui/commit/8b1976c6e35833ec02ee15deba440454cc888283 there should not be issues with long passwords.

**Comment by j4ypeem** [comment]
Last night I replaced the GUI with the c354c1 build. This morning I've encountered the GUI crash again. Are there any GUI logs that I can gather that will help troubleshoot?

**Comment by lstipakov** [comment]
Good, this is expected. We're narrowing down the changelog to find an offensive commit. 

Those are remaining suspects:

```
PS C:\Users\lev\Projects\openvpn-build\src\openvpn-gui> git bisect view --oneline
c354c19 .gitignore: update
53ae284 Update CHANGES.rst
28a5682 Handle CONNECTED,ROUTE_ERROR state message
b4e3e07 Support colouring of WriteStatusLog messages
d47b5f1 Update italian language
b25a24f Update italian language
f7d594d Update italian language
99a0c5d Fix truncated text in Italian
fa82d69 (HEAD) Update Italian language
cea3d25 Update openvpn-gui-res-zh-hans.rc
6398b1d (tag: v11.34.0.0, tag: OpenVPN-2.6_rc1-I001) Bump version to 11.34.0.0
6daee58 Update CHANGES.rst
548aeb0 Improve README
442a48e BUILD.rst: switch to VS2022
e9e9dae Update vcpkg commit id
```

Please try this one. 
[openvpn-gui-fa82d6.zip](https://github.com/OpenVPN/openvpn-gui/files/12417406/openvpn-gui-fa82d6.zip)

There might be dumps at `%LOCALAPPDATA%\CrashDumps`.

**Comment by j4ypeem** [comment]
Will try this one and report back. Probably tomorrow.
Thanks

**Comment by j4ypeem** [comment]
Same happened with this build but no crash dump files again.

**Comment by lstipakov** [comment]
Thanks, how about 
[openvpn-gui-548aeb.zip](https://github.com/OpenVPN/openvpn-gui/files/12418744/openvpn-gui-548aeb.zip) ?

**Comment by j4ypeem** [comment]
Same crash happens with this one. No crash dump files.

**Comment by lstipakov** [comment]
Hm.. than doesn't make much sense, the rest of commits do not have any code changes. Let's make sure that our baseline is correct. Can you retest the latest one considered as "good"? I have rebuilt it. 
[openvpn-gui-8b1976c-v2.zip](https://github.com/OpenVPN/openvpn-gui/files/12419585/openvpn-gui-8b1976c-v2.zip)

**Comment by lstipakov** [comment]
That “vcpkg commit-id” related to GitHub Actions only (the commit message could be more clear). Lähetetty iPhonestaSelva Nair ***@***.***> kirjoitti 23.8.2023 kello 17.39:﻿
If 8b1976c is indeed well-behaved, this points to dependency library updates (vcpkg commit-id change) as the culprit?

**Comment by selvanair** [comment]
> That “vcpkg commit-id” related to GitHub Actions only

I see, so all the git-bisect builds you have provided are linked to the same set of dlls..

For memory dumps the user may have to enable it by creating a registry key: HKLM\SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps\openvpn-gui.exe

**Comment by chipitsine** [comment]
> Same crash happens with this one. No crash dump files.

please check whether Windows is configured to save dumps.

https://learn.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps

looks like those settings are missing. having dumps usually helps a lot

**Comment by lstipakov** [comment]
Crash dumps are already enabled during installation starting from from [2.6.5](https://github.com/OpenVPN/openvpn-build/commit/b72896baa11a7845944622fe68f9d3786e914905).

**Comment by j4ypeem** [comment]
I've checked and the following reg keys are already present:
Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps\openvpn.exe
Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps\openvpn-gui.exe
Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\Windows Error Reporting\LocalDumps\openvpnserv.exe

But there were no values configured - such as DumpFolder, DumpCount, DumpType or CustomDumpFlags.

![image](https://github.com/OpenVPN/openvpn-gui/assets/127693073/c8fe5b56-7559-4c21-b7ae-80f4997907de)

From reading the Microsoft documentation linked above, the specific values should be present.

![image](https://github.com/OpenVPN/openvpn-gui/assets/127693073/cd1f81f3-8e3b-4e9e-9476-aeb976b63636)

![image](https://github.com/OpenVPN/openvpn-gui/assets/127693073/70b9ac4b-f278-4dd4-8aff-f6579935271f)

I've manually created the default values from that linked document.

![image](https://github.com/OpenVPN/openvpn-gui/assets/127693073/a2d8e8de-66e0-49b0-a865-e8209abd5dd1)

I'm not sure what build version you wanted me to try next...so I've reverted back to the original version before I was sent custom builds via this thread.

Hopefully these changes will give us some crash dumps for analysis.

**Comment by j4ypeem** [comment]
Adding those registry values didn't make any difference. I tried the following again this morning and no crash dumps for either:

Original version - GUI crashed and no crash dumps
Build c354c1 - GUI crashed and no crash dumps

**Comment by lstipakov** [comment]
Not sure those values must be added, since documentation explicitly mentions default values.

Let's try to find the last version that works. Can you retest with the build from [this](https://github.com/OpenVPN/openvpn-gui/issues/548#issuecomment-1689954992) comment?

**Comment by j4ypeem** [comment]
Tried again with openvpn-gui-8b1976c-v2.zip and same issue

**Comment by lstipakov** [comment]
The same issue - crash and no dumps?

**Comment by j4ypeem** [comment]
yes - crash and no dumps

**Comment by lstipakov** [comment]
This [openvpn-gui-8f3c7a0.zip](https://github.com/OpenVPN/openvpn-gui/files/12429854/openvpn-gui-8f3c7a0.zip) is a fairly old version (2.5.7, May 2022), please give it a try.

**Comment by j4ypeem** [comment]
I'm now using build 8f3c7a0. So far so good - I'll continue to monitor for the rest of the day and tomorrow.
FYI - I am available for further testing tomorrow (Friday) but after then, I'm unlikely to be available to assist with further troubleshooting for a number of weeks.

**Comment by j4ypeem** [comment]
Using build 8f3c7a0, I just disconnected from one of my active VPN's and the GUI crashed. No crash dumps again.

**Comment by lstipakov** [comment]
Okay, so the bug seems to be an old one. Could you please give a try to 
[openvpn-gui-6758dd.zip](https://github.com/OpenVPN/openvpn-gui/files/12436350/openvpn-gui-6758dd.zip) which is 2.5.4 from October, 2021?

**Comment by j4ypeem** [comment]
Build openvpn-gui-6758dd gave me the same issue.... crashed when disconnecting from one of two active VPN connections.

I then reverted to build version 8b1976c... which was the first custom build that you shared. This one was originally running fine for a few days, but it's just crashed for me now. It crashed when connecting to a second VPN. The GUI icon disappeared, but the connection did successfully establish.

We seem to be going round in circles a bit here.... I think it would be worth trying to investigate and resolve why no CrashDumps are being generated. Without these.... we're running blind and have no visibility into the specific root cause.

I've got these registry settings configured:

![image](https://github.com/OpenVPN/openvpn-gui/assets/127693073/d8d45fa0-9881-44f6-aae4-376a158a998c)

![image](https://github.com/OpenVPN/openvpn-gui/assets/127693073/4b08aa57-696b-473c-bf0f-91704306bce5)

My understanding is that the more specific settings of the openvpn-gui.exe key should take preference. Regardless, there is nothing in either the C:\TEMP\CrashDumps or %LOCALAPPDATA%\CrashDumps folders.

Can you confirm that the openvpn-gui.exe runs in the context of the "END_USER" account - so the full path of %LOCALAPPDATA%\CrashDumps should be C:\Users\END_USER\AppData\Local\CrashDumps

**Comment by j4ypeem** [comment]
As mentioned in this comment - https://github.com/OpenVPN/openvpn-gui/issues/548#issuecomment-1691741173, I'm unavailable to assist with troubleshooting after today, My colleague @sscotter might be able to continue these investigations in the meantime.

**Comment by chipitsine** [comment]
I was curious whether openvpn installer really configures crash dumps or not.

so, I installed recent openvpn msi (it configured registry), I added `int i = 10, j = 0;  i = i / j;` to _twinMain...

and there're dumps

![image](https://github.com/OpenVPN/openvpn-gui/assets/2217296/0abcef07-a0ce-4f26-a6d2-21e1c0849773)

**Comment by chipitsine** [comment]
maybe some other registry settings overrides crash dump behaviour ? I am curious what `gpresult /H gp.html` will show

**Comment by lstipakov** [comment]
So I just enabled ASAN (took a while to figure out that it is not compatible with /Qspectre) and got a crash when starting the second connection. Looking into that.

![asan](https://github.com/OpenVPN/openvpn-gui/assets/643110/7f62cacb-e549-4556-bf1e-ae001b8debfb)

**Comment by j4ypeem** [comment]
> So I just enabled ASAN (took a while to figure out that it is not compatible with /Qspectre) and got a crash when starting the second connection. Looking into that.
> 
>

Did you find anything?

I'm still having the same problems - I'm currently using one of the custom builds that were provided earlier in this thread.
I've got some crashdumps and have attached one here. Hopefully this can help in the troubleshooting process.

Thanks

[openvpn-gui.exe.19752.zip](https://github.com/OpenVPN/openvpn-gui/files/12828946/openvpn-gui.exe.19752.zip)

**Comment by lstipakov** [comment]
Unfortunately I have not found anything and ASAN exception seem to be unrelated yo problem you experience. I still believe that the best course of actions would be to find the last working version and then bisect to the problematic commit. 

Is it so that 2.5.4 also crashes for you?

**Comment by j4ypeem** [comment]
> Unfortunately I have not found anything and ASAN exception seem to be unrelated yo problem you experience. I still believe that the best course of actions would be to find the last working version and then bisect to the problematic commit.
> 
> Is it so that 2.5.4 also crashes for you?

I have reverted to 2.5.4 for testing purposes. Will keep you updated.

**Comment by j4ypeem** [comment]
Have just experienced the same issue of the disappearing GUI.
Crashdump attached
![image](https://github.com/OpenVPN/openvpn-gui/assets/127693073/05064761-1f1b-4d16-9931-e3e4d7101c5d)
[openvpn-gui.exe.15680.zip](https://github.com/OpenVPN/openvpn-gui/files/12866933/openvpn-gui.exe.15680.zip)

**Comment by j4ypeem** [comment]
@lstipakov Here's another crashdump if it helps?
[openvpn-gui.exe.8904.zip](https://github.com/OpenVPN/openvpn-gui/files/12867220/openvpn-gui.exe.8904.zip)

**Comment by lstipakov** [comment]
I checked the first dump and (unsurprisingly) again heap corruption. The stack trace points to the code from 2017:

```
0:000> !analyze -v
*******************************************************************************
*                                                                             *
*                        Exception Analysis                                   *
*                                                                             *
*******************************************************************************

*** WARNING: Unable to verify timestamp for openvpn-gui.exe

KEY_VALUES_STRING: 1

    Key  : Analysis.CPU.mSec
    Value: 31

    Key  : Analysis.DebugAnalysisManager
    Value: Create

    Key  : Analysis.Elapsed.mSec
    Value: 12539

    Key  : Analysis.Init.CPU.mSec
    Value: 218

    Key  : Analysis.Init.Elapsed.mSec
    Value: 10223

    Key  : Analysis.Memory.CommitPeak.Mb
    Value: 100

    Key  : Timeline.OS.Boot.DeltaSec
    Value: 216

    Key  : Timeline.Process.Start.DeltaSec
    Value: 156

    Key  : WER.OS.Branch
    Value: ni_release

    Key  : WER.OS.Timestamp
    Value: 2022-05-06T12:50:00Z

    Key  : WER.OS.Version
    Value: 10.0.22621.1

    Key  : WER.Process.Version
    Value: 11.25.0.0


FILE_IN_CAB:  openvpn-gui.exe.15680.dmp

NTGLOBALFLAG:  0

PROCESS_BAM_CURRENT_THROTTLED: 0

PROCESS_BAM_PREVIOUS_THROTTLED: 0

APPLICATION_VERIFIER_FLAGS:  0

CONTEXT:  (.ecxr)
rax=000023183d311507 rbx=00000000c0000374 rcx=00000000000004b4
rdx=0000000000000003 rsi=0000000000000001 rdi=00007ffcb0b718a0
rip=00007ffcb0afc319 rsp=00000018054fcac0 rbp=000001671a2ef7e0
 r8=00007ffcb0a0d061  r9=0000000000000002 r10=00000018054fbe20
r11=00000018054fbe20 r12=0000000000000000 r13=000001671a210000
r14=0000000000000000 r15=0000000000000000
iopl=0         nv up ei pl nz na pe nc
cs=0033  ss=002b  ds=002b  es=002b  fs=0053  gs=002b             efl=00000202
ntdll!RtlReportFatalFailure+0x9:
00007ffc`b0afc319 eb00            jmp     ntdll!RtlReportFatalFailure+0xb (00007ffc`b0afc31b)
Resetting default scope

EXCEPTION_RECORD:  (.exr -1)
ExceptionAddress: 00007ffcb0afc319 (ntdll!RtlReportFatalFailure+0x0000000000000009)
   ExceptionCode: c0000374
  ExceptionFlags: 00000081
NumberParameters: 1
   Parameter[0]: 00007ffcb0b718a0

PROCESS_NAME:  openvpn-gui.exe

ERROR_CODE: (NTSTATUS) 0xc0000374 - Keko on vioittunut.

EXCEPTION_CODE_STR:  c0000374

EXCEPTION_PARAMETER1:  00007ffcb0b718a0

STACK_TEXT:  
00000018`054fcac0 00007ffc`b0afc2e3     : 00000018`05500000 00007ffc`b0b718f8 00007ffc`b0b7a000 00007ffc`b09f0000 : ntdll!RtlReportFatalFailure+0x9
00000018`054fcb10 00007ffc`b0b0537a     : 00000000`00000002 00007ffc`b0b718a0 00000000`00000003 00000167`1a210000 : ntdll!RtlReportCriticalFailure+0x97
00000018`054fcc00 00007ffc`b0b0565a     : 00000000`00000003 00000167`1a2ef7e0 00000167`1a210000 000058bc`363851ff : ntdll!RtlpHeapHandleError+0x12
00000018`054fcc30 00007ffc`b0b11655     : 00000167`1a2ef7ef 00000167`1a2ef7f0 00000018`054fcd68 00000167`1a2874b8 : ntdll!RtlpHpHeapHandleError+0x7a
00000018`054fcc60 00007ffc`b0a9be18     : 00000000`00000000 00000167`1a2874b8 00000167`1a2874b8 00000018`054fcd68 : ntdll!RtlpLogHeapFailure+0x45
00000018`054fcc90 00007ffc`b0a2ab01     : 00000167`1a2ef7e0 00000167`1a210000 00007ffc`a62962c0 00000000`00000000 : ntdll!RtlpFreeHeapInternal+0x707a8
00000018`054fcd50 00007ffc`b081668d     : 00000000`00000000 00000018`054fce08 00000018`054fce08 00000167`1a2efb50 : ntdll!RtlFreeHeap+0x51
00000018`054fcd90 00007ffc`a6282fc5     : 00007ffc`a6286f01 00000000`00000000 00000000`00000000 00000000`00000000 : rpcrt4!RpcBindingFromStringBindingW+0xed
00000018`054fcdc0 00007ffc`b083685b     : 00000000`00000000 00007ffc`a6286f60 00000000`00000000 00000000`00000000 : samlib!PSAMPR_SERVER_NAME_bind+0xb5
00000018`054fce40 00007ffc`b083676a     : 00000018`054fd200 00000000`00000040 00007ffc`a6296000 00000000`00000000 : rpcrt4!GenericHandleMgr+0xa7
00000018`054fce70 00007ffc`b08b260e     : 00000000`00000001 00000000`00000000 00000018`054fcf40 00000000`00000001 : rpcrt4!ExplicitBindHandleMgr+0x7a
00000018`054fceb0 00007ffc`b08b1b6b     : 00007ffc`a6296220 00000000`00000040 00000000`00000000 00007ffc`a6296000 : rpcrt4!NdrpClientCall3+0x5ce
00000018`054fd1d0 00007ffc`a62827f0     : 00000167`00000040 00000018`054fd8e0 00000000`00000000 00000000`00000000 : rpcrt4!NdrClientCall3+0xeb
00000018`054fd560 00007ffc`a5d22483     : 00000167`00000000 00000000`00000000 00000000`00000000 00000167`1a2daa20 : samlib!SamConnect+0x120
00000018`054fd870 00007ff7`1dd518cd     : 00000000`0000000c 00007ffc`b0a2be60 00000000`0000000d 00007ffc`b0a2be60 : samcli!NetLocalGroupGetMembers+0x143
00000018`054fd9f0 00000000`0000000c     : 00007ffc`b0a2be60 00000000`0000000d 00007ffc`b0a2be60 00000018`ffffffff : openvpn_gui+0x18cd
00000018`054fd9f8 00007ffc`b0a2be60     : 00000000`0000000d 00007ffc`b0a2be60 00000018`ffffffff 00000018`054fda30 : 0xc
00000018`054fda00 00007ffc`b0a2ab01     : 00000167`1a2a4c80 00000167`1a210000 00000167`1a212716 00000000`00000005 : ntdll!RtlpFreeHeapInternal+0x7f0
00000018`054fdac0 00007ff7`1dd727ec     : 00000167`1a2a5350 00007ffc`b0a2d19a 00000000`002f001e 00000000`00000040 : ntdll!RtlFreeHeap+0x51
00000018`054fdb00 00000167`1a2a5350     : 00007ffc`b0a2d19a 00000000`002f001e 00000000`00000040 00000167`05000b0e : openvpn_gui+0x227ec
00000018`054fdb08 00007ffc`b0a2d19a     : 00000000`002f001e 00000000`00000040 00000167`05000b0e 00000000`00000000 : 0x00000167`1a2a5350
00000018`054fdb10 00000000`00000000     : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : ntdll!RtlpLowFragHeapAllocFromContext+0x1ba


STACK_COMMAND:  ~0s; .ecxr ; kb

SYMBOL_NAME:  samlib!PSAMPR_SERVER_NAME_bind+b5

MODULE_NAME: samlib

IMAGE_NAME:  samlib.dll

FAILURE_BUCKET_ID:  HEAP_CORRUPTION_c0000374_samlib.dll!PSAMPR_SERVER_NAME_bind

OS_VERSION:  10.0.22621.1

BUILDLAB_STR:  ni_release

OSPLATFORM_TYPE:  x64

OSNAME:  Windows 10

IMAGE_VERSION:  10.0.22621.2283

FAILURE_ID_HASH:  {ac756c3e-08bb-4ef2-64bd-f70168d3c4a1}

Followup:     MachineOwner

```

The second dump shows completely different location.

Looks like we need to go further backwards. Could you try 2.5.0?

**Comment by j4ypeem** [comment]
I reverted to OpenVPN-2.5.0-I601-amd64.msi
Had the same issues
![image](https://github.com/OpenVPN/openvpn-gui/assets/127693073/e868edf7-16ce-4a30-bb09-9bfeaf051046)

2 crashdumps attached

[openvpn-gui.exe.17384.zip](https://github.com/OpenVPN/openvpn-gui/files/12878912/openvpn-gui.exe.17384.zip)

**Comment by opoplawski** [comment]
With a new Windows 11 laptop and OpenVPN 2.6.8-I001 (gui 11.46.0.0) one of our users is hitting this hard.  Haven't really seen it before.  Not sure what would be different about his setup.

**Comment by j4ypeem** [comment]
@lstipakov Are you free to continue investigating this please?
It seems that @opoplawski is having similar issues?

**Comment by opoplawski** [comment]
After an initial flurry of crashes, it's settled down a lot - but there were still two crashes on the 6th.

**Comment by mduivcap** [comment]
I still have the crash on a daily basis. To me it really seems like a timing issue.

- When the password is wrong, it never crashes, it just tells me the password is wrong, so up to that point everything seems to go well
- Sometimes it does not crash and it connects as expected 
- Sometimes it crashes but did connect underwater
- More often it crashes and I need to start up the application again and try again
- And once every so many attempts it seems like it's not connected, but does occupy the TAP adapter

**Comment by Pjstolas** [comment]
Just to let you know, i'm having the same behavior, connection crashes the Ovpn gui and connections goes through. Win11 22H2.

EDIT: And suddenly connection time crashed... oh god.

**Comment by mduivcap** [comment]
The workaround I'm using now is to use the CLI instead of the GUI.  

1. Create a .bat file with the following command
```cmd
@echo off
"C:\Program Files\OpenVPN\bin\openvpn.exe" --config "C:\<whatever-location-you-have-stored-your-config>\Config.ovpn"
pause
```
2. Create a shortcut on the desktop (with admin permissions) to run the bat file

**Comment by Kajetan321** [comment]
We  have the same problem, for a few users the OpenVPN GUI crashes randomly (mousing over the icon makes the icon disappear).  I noticed however that this is configuration/server dependent.  The problem comes up when connection for our new pfSense firewall.  Using the same client that crashed with pfSense to connect to an old Ubiquity router is rock solid.  For what it's worth here is the pfSense configuration and then the Ubiquity configuration.

`dev tun
persist-tun
persist-key
data-ciphers AES-256-GCM:AES-128-GCM:CHACHA20-POLY1305:AES-256-CBC
data-ciphers-fallback AES-256-CBC
auth SHA256
tls-client
client
resolv-retry infinite
remote xx.xxx.xx.xx xxxx udp4
nobind
verify-x509-name "OpenVPN Server Cert" name
auth-user-pass
remote-cert-tls server
explicit-exit-notify
auth-nocache`

`client
dev tun
proto udp
remote xxx.xxx.com xxxx
float
resolv-retry infinite 
nobind
persist-key 
persist-tun 
remote-cert-tls server
auth-nocache
verb 3
ca cacert.pem 
cert xxxxx.pem
key xxxxx.key`

**Comment by Kajetan321** [comment]
I think I figured out what's causing the issue.  The problem seems to be caused by special characters in the passwords. My passwords were generated randomly and included special characters.  I regenerated the passwords for the two offending users without special characters and behold the problem stopped occurring.  I haven't had the time to figure out which character is causing the problem but here are the two passwords that appear to trigger the bug:

TFR.whGN>8"/gQzaE[U*m]
 
Y@Ps['pj)Na7JvT$cQ2k="

Both passwords include the quotation mark.  Could that be it?

**Comment by selvanair** [comment]
@Kajetan321 Amazing find. 

We do have an off-by-one  error in the special character (",\ and space) escape function:

https://github.com/OpenVPN/openvpn-gui/blob/9c0c48889082c92ef17b2ab221a2a26f2c4544af/misc.c#L211
            
`memmove(out + pos + 1, out + pos, len - pos + 1);`

`out` has space only for `len + 1` bytes and we should be copying only `len - pos` characters.

Looks like the bug was always there (at least since 2012: https://github.com/OpenVPN/openvpn-gui/commit/976b0831). Later I copy pasted it without checking the logic.

@Kajetan321  Could you please check whether this binary fixes it ?
https://github.com/selvanair/openvpn-gui/actions/runs/8712771712#artifacts
Download the zip file for your platform and you will find openvn-gui.exe in there. Copy the executable to a temp location, stop the GUI and run this one. No installation required.

**Comment by Alexey-Kovalenko** [comment]
Hmm, when GUI crashes,  my passwords has space, comma and <

**Comment by selvanair** [comment]
> Hmm, when GUI crashes, my passwords has space, comma and <

Space is also a character we need to escape and will exercise that buggy function. The binary linked above should fix it.

**Comment by Alexey-Kovalenko** [comment]
> > Hmm, when GUI crashes, my passwords has space, comma and <
> 
> Space is also a character we need to escape and will exercise that buggy function. The binary linked above should fix it.

Not fixed :(
I'm tried password of 16 spaces, and  GUI crashes near every connect. 8 spaces - rare. 24 spaces - every time.

**Comment by lstipakov** [comment]
@Alexey-Kovalenko  Does it not crash if you have a password which doesn't need escaping?

**Comment by lstipakov** [comment]
I have [refactored](https://github.com/lstipakov/openvpn-gui/commit/7746be31ca497d46216ad4ee72c92392e2e0b049) `escape_string()`, could you give it a try? [https://github.com/lstipakov/openvpn-gui/actions/runs/8752747296/artifacts/1429277978](https://github.com/lstipakov/openvpn-gui/actions/runs/8752747296/artifacts/1429277978)

**Comment by selvanair** [comment]
@Alexey-Kovalenko Are you sure you are running the binary from the link I had posted? Infact I'm unable to run that binary because of a missing dll -- it seems to have a dependency on clang_rt.asan dll

Try the following from the latest build
https://github.com/OpenVPN/openvpn-gui/actions/runs/8735256061/artifacts/1425177408
This one is with ovpn3 support and depends on json dll which is included.

@lstipakov Any idea why x64 artifact is sometimes picked from the asan build? The refactor build you posted has non-asan binary  in the artifacts! Is one over-writing the other?

**Comment by Kajetan321** [comment]
@lstipakov I tried yours and  @selvanair zip files, they both are asking for a .dll.  Yours is looking for vcruntime140.dll, selvanair's wanted a different dll.

**Comment by selvanair** [comment]
@Kajetan321 Sorry about that -- something is up with our CI builds.
Could you try the exe in the link below please?
https://github.com/selvanair/openvpn-gui/releases/tag/memmove_fix
Its an executable locally built using mingw. There should be no dependencies.

**Comment by Alexey-Kovalenko** [comment]
@selvanair Yes, It was openvpn-gui_x64_Object.zip, inside 
libopenvpn_plap.dll 17.04.2024 16:12:16
openvpn-gui.exe17.04.2024 16:12:16
test_plap.exe 17.04.2024 16:12:16
I put gui ans dll to openvpn bin catalog.

Good new, everyone.
GUI from latest build is working even with 32 spaces password.

**Comment by Alexey-Kovalenko** [comment]
> @Alexey-Kovalenko Does it not crash if you have a password which doesn't need escaping?

Well, no. I remove space from password, and no crashing.

**Comment by Alexey-Kovalenko** [comment]
Extra question - can OpenVPN GUI work win non-english passwords? When character code is greater then 127?
I think, GUI works with 2-byte unicode characters, but console openssl works with 1-byte codepages, so it is hard to convert :(

**Comment by selvanair** [comment]
> Extra question - can OpenVPN GUI work win non-english passwords? When character code is greater then 127?
> I think, GUI works with 2-byte unicode characters, but console openssl works with 1-byte codepages, so it is hard to convert :(

The GUI gets the password in UTF-16 from the dialog, converts into UTF-8 and passes to OpenVPN core. The latter treats the password as a utf8 string, so it should all work correctly provided what you have in mind as your password matches that utf16->utf8 conversion.  The only character not allowed is "\n" (newline). However, you are also dependent on how a verification script or plugin at the server side would interpret it. 

That said, as a general rule, I would stick to printable ascii characters for all passwords.

**Comment by boschhd** [comment]
I have the same issue with the gui crashing and the error code

> Faulting application name: openvpn-gui.exe, version: 11.44.0.0, time stamp: 0x64db3fb6
> Faulting module name: ntdll.dll, version: 10.0.19041.4355, time stamp: 0x35a939aa
> Exception code: 0xc0000374

One user mentioned this line as relevant:

> SendMessage(logWnd, EM_SCROLLCARET, 0, 0);

I disabled the "show ballon" option in the gui settings and it stopped crashing for me now :) 
Maybe this helps

**Comment by selvanair** [comment]
The issue discussed in this thread was caused by blanks in password and has been fixed. Should be in next release or you can use openvpn-gui.exe from CI artifacts to test: https://github.com/OpenVPN/openvpn-gui/actions/runs/8775405220#artifacts

if you are seeing crash unrelated to having space, backslash or quote characters in password or username, please open a new issue.
